### ISSUE TITLE
[DSIG] Mistake in Format 1:, d: delete "Zero out the file checksum in the 'head' table"

### ISSUE THREAD

**Issue by ValZapod** [MicrosoftDocs/typography-issues#455]
You can look in this thread, for example. https://lists.aau.at/pipermail/mpeg-otspec/2016-December/000837.html

> a) update the spec to match the implementation (i.e. remove the line "Zero
> out the file checksum in the head table.", as it is *not* actually
> performed by signcode.exe + mssipotf.dll, as I and Hin-Tak have repeatedly
> explained);

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: acf0601c-2dd5-c94c-8860-07e9def0c20c
* Version Independent ID: c69a8340-3140-31cf-d27b-c0e2b235ccc6
* Content: [DSIG — Digital signature table specification - Typography](https://docs.microsoft.com/en-us/typography/opentype/spec/dsig)
* Content Source: [typographydocs/opentype/spec/dsig.md](https://github.com/MicrosoftDocs/typography/blob/live/typographydocs/opentype/spec/dsig.md)
* Product: **typography**
* GitHub Login: @PeterCon
* Microsoft Alias: **PeterCon**

**Comment by alib-ms** [comment]
Thank you for the feedback. This issue will be triage and addressed in the version of the OpenType specification.

**Comment by ValZapod** [comment]
@PeterCon It would be nice to understand what is the right thing here because "Zero
out the file checksum" should be done, no? Or is Authenticode not caunting on that part of binary that is that checksum?

**Comment by PeterCon** [comment]
@ValZapod : As you point out, this issue has been lingering for some time. I will endeavour to find some resolution to this soon. It may require getting input from several platform or tool vendors.

**Comment by ValZapod** [comment]
Or we can just do @HinTak, please could you comment how to word it so we can close it ASAP?

**Comment by HinTak** [comment]
According to my notes and slides I gave a talk on 2016, the current official signing implementation (ie Microsoft's) does not zero out the checksum. So to adjust the spec to the De facto implementation, the 4th item "zero out the file checksum..." should be removed. (with perhaps a note saying it used to be mistakenly there).

About ttc header, the bulk of many old Microsoft shipped ttc also have a version 1 header rather than version 2. So that's another issue where the De facto signing implementation (does not reject v1 header) does not match the spec (which only states v2).

**Comment by ValZapod** [comment]
> does not zero out the checksum

So it does not use those bytes of checksum in its calculations of RSA md5 signature?? How does it work? Because if it is not zeroed, and then rewritten, there is no way to check the signature as the file was modified (with the new checksum).

**Comment by HinTak** [comment]
Huh? The checksums are per table. The signature is based on all the tables plus checksum fields (with valid content), rather than all the tables minus the checksum fields (with zero'ed content).

Thus the signature also certifies the checksums, rather than excluding them. Clear?

**Comment by ValZapod** [comment]
> on all the tables

No, only head table. From spec:

>Zero out the file checksum in the 'head' table

Does DSIG table itself has a checksum and if yes, how it is done?

**Comment by HinTak** [comment]
You need to read the spec - the head table stores the checksum of **all the other** tables.

**Comment by ValZapod** [comment]
> **all the other** tables

Including DSIG? But then it means checksum of checksums that is head table should be **AGAIN** recalculated and zeroed. Again, what it is used to calculate md5?

**Comment by HinTak** [comment]
No, excluding DSIG - since it did not exist before signing. You can read the code in the DSIGInfo directory in FontVal (which contains an implementation of verifying signature), if you don't understand the spec.

**Comment by ValZapod** [comment]
So it also does not "Recalculate the checksum in the 'head' table."? Can you give a pull request maybe.

**Comment by HinTak** [comment]
?? @ValZapod  Look: it was discussed at quite some length 4 years ago. I think the people involved understand the issue well enough. If you want to spend time on it, by all means, please do so quietly, and please stop making demand on other people's time on an hourly basis.

**Comment by ValZapod** [comment]
> demand on other people's time on an hourly basis.

Haha, well, you did not manage to modify the Microsoft docs in these 4 years or do an opensource implemenation and otf still have md5 hashes inside, pathetic, everybody is usg sha2 already. Again, you either zero out and recalculate or not zero out and not recalculate. Or 3, somehow md5 digest is not counting that checksum. What is it?

**Comment by HinTak** [comment]
@PeterCon was in on the discussion 4 years ago, and he is on this now. I have nothing to add beyond what I already wrote, and therefore prefer not to respond to this thread.

**Comment by PeterCon** [comment]
@HinTak : Can you remind me where this was discussed 4 years ago? (Sorry, I was removed from work in this area two years ago and have lost track of such details.)

**Comment by HinTak** [comment]
@PeterCon the beginning of this post seems to be part of a series of exchanges from an archive of the mpeg otspec formerly on yahoo lists. Somebody from Microsoft (at the bottom of https://lists.aau.at/pipermail/mpeg-otspec/2016-December/000837.html , and I believe more than one person too) did respend but it did not go anywhere further.

As I wrote earlier - 
https://github.com/MicrosoftDocs/typography-issues/issues/455#issuecomment-679192579 - from my slides of a talk I gave in 2016, I would suggest removing the 4th clause "zero out the file checksum..." of the item group of 5, as this is how the actual de facto implementation behaves; also the spec says a v2 ttc header contains fields for DSIG, but many older ttc shipped by Microsoft does not do this (i.e. the ttc has a v2-like header where the version field is 1 instead of 2). The latter is more a laxness in the signing tool not refusing to sign a non-conformant font. So both of these issues are about the spec not agreeing with the signing tool.

However, as you know, the de facto verification implementation is also in that microsoft signing dll, it accepts what it signs :-).

**Comment by PeterCon** [comment]
I've marked this as a P1 issue and have asked MS to provide clarification.

**Comment by HinTak** [comment]
@PeterCon yes, I think that the proper course of action: find somebody within Microsoft who has access to the source code of the signing/verifying opentype authenticode dll, have a look and verify this claim, and either fix the signing/verifying tool or fix the spec. At this stage, after so many fonts have been signed for almost two decades, I think it is the spec that needs to be fixed. However, the first step (hunt down the source within Microsoft and get a tech opinion on it) must be done by a Microsoft staffer - outside parties can only keep poking... And I have done my share of poking a few years ago so am a bit tired of poking MS people every few months ...

**Comment by behdad** [comment]
Can't we just deprecate `DSIG` table completely and mark it for future removal?

**Comment by ValZapod** [comment]
> Can't we just deprecate `DSIG` table completely

Does the word stable API means anything to you? Wow. DSIG, even incorrect, is what is used to activate otf styles sets or variables or ligartures. Also we can only used signed files in bootloader. And we should use everything signed with SHA2.

**Comment by PeterCon** [comment]
@ValZapod I think this is valuable information. Can you clarify the applications or environments in which DSIG is a factor (i) for activating OT features or (ii) for a bootloader?

And when you say, "I always check" do you mean when you as an individual user download fonts from a site, or are you referring to some software product that retrieves and uses fonts from sites? And what is being used to verify the signature in this scenario?

**Comment by ValZapod** [comment]
BTW, I just checked, ISO/IEC 14496-22:2015(E) -- not 2019, I do not have it -- also says that `Zero out the file checksum in the head table` thing. You need to give that data to ISO/IEC as well, haha.

**Comment by ValZapod** [comment]
> individual user download fonts from a site

No, when I download pirated fonts packs from our ru-board forum... There are other forums of course. For example, DG META SCIENCE (ala ITC Officina Super family) or my favourites non-free from Latex https://tex.stackexchange.com/questions/425098/which-opentype-math-fonts-are-available

> for a bootloader?

Is not Windows Secure Boot EUFI bootloader uses ONLY signed fonts? And not with just a dummy signature. It is not like that in debian, it was discussed that this should be fixed. There is after all some bytecode in otfs as well.

> for activating OT features

Hard to say. It is required in standards to activate OTF, so who knows, you can strip DSIG from font and play with it. Word does not care, for example. Adobe though, is prety sensitive about it, especially Illustrator.

**Comment by Gr3gH** [comment]
I’ll take a look at this, but it may be sometime before I can get back with a response.

**Comment by behdad** [comment]
> Does the word "stable API" mean anything to you?

Yes. I've been maintaining various C libraries that provide ABI/API compatibility indefinitely. I've maintained them for over a decade.

> Wow. DSIG, even incorrect, is what is used to activate otf styles sets or variables or ligartures.

That's irrelevant. If a stupid piece of software does that, that's a software bug.

> Also we can only use signed files in bootloader. And we should use everything signed with SHA2. Ah, yes, do you wanna know how many vulns are there because of malformed otfs? Hundreds. Also, when I download fonts from some less legal places, I always check that signature is correct.

That's redherring. Just sign the font files externally using generic signing tools like you sign your kernel image or any other files. I'm sure your bootloader does not load and verify DSIG.

**Comment by ValZapod** [comment]
> kernel image

Is a ELF. It has special place for signature and timestamp signature. PE is even more compicated with 2 styles of signatures and timestamp signatures. PDF is again different. Doc and docx is again different, both of them. 
> stupid piece of software does that, that's a software bug

That is what ISO/IEC standards say. They do not ask to verify signature though or even check if it is there though.
Here you are https://typedrawers.com/discussion/1147/are-empty-dsig-tables-in-a-font-useless

**Comment by vlevantovsky** [comment]
> DSIG, even incorrect, is what is used to activate otf styles sets or variables or ligartures. 

Gong way back to initial WOFF2 discussions, our decision about removing DSIG from WOFF2 files was partially informed by this message from Sergey Malkin: https://lists.w3.org/Archives/Public/public-webfonts-wg/2014May/0025.html

**Comment by paullinnerud** [comment]
I looked at the code for mssipotf.dll. The code looks pretty clear both in comments and code that the entire file is hashed to come up with the signature. The same hash function is used in both signing and verification. In the simplest case where the input font has no DSIG table the entire file input file is mapped and the hash is created. In the case where there is an existing DSIG table, a new internal font is created with the DSIG removed with the intention of creating exactly the same as original file so that can be hashed and eventually compared with an encrypted hash in the DSIG table in the case of verification or when signing that internal image has a new signature added based on the new hash.

**Comment by ValZapod** [comment]
> The same hash function is used in both signing and verification.

It is literally impossible. Signing is rsa signature of md5 digest. And to verify you need to check the whole chain of 3 certificates.

> entire file is hashed

We know that. The question is how the entire file is hashed.

**Comment by paullinnerud** [comment]
The same internal function in mssipotf.dll is used to create the hash in both signing and verification of font files. The simplest case is when the font does not have an existing DSIG to be verified or to be replaced. This case only occurs when you have a previously unsigned font to be signed. 

In this simplest case the entire file is passed into CryptHashData() https://docs.microsoft.com/en-us/windows/win32/api/wincrypt/nf-wincrypt-crypthashdata. 

The SIP is built on an infrastructure in Microsoft Windows where various file types can be signed. There is build in support for many file types including .exe, .dll, etc. Font files were not originally built into the code so mssipotf.dll was created to extend the signing infrastructure to support .ttf and .ttc files.

**Comment by ValZapod** [comment]
CryptHashData() and CryptCreateHash() is just doing an md5. You will need to then sign with RSA that md5. And then maybe get a timestamp from timestamp server from Verisign, dunno (?). It is complicated. Now the point here is that the question is about The checksums of all tables are correct.

**Comment by paullinnerud** [comment]
It does not appear that the code zeros out the checksum in the 'head' table prior to calling CryptHashData(). Once the hash is created and encrypted and added to the DSIG structure and added to the font then the checksum in the head table is recalculated. The reason this does not break the signature is when the signature is verified, an in memory image is created of the font without the DSIG and the checksum is recalculated before the data is hashed. 

Prior to signing, the checksums are verified to make sure they are correct. Signing will be refused if those checks are not successful.

**Comment by ValZapod** [comment]
> an in memory image is created of the font without the DSIG and the checksum is recalculated before the data is hashed

THANK YOU! What about usFlag? Is it added afyer or before the checksum is recalculated in memory image?

**Comment by paullinnerud** [comment]
Are you referring bit 11 of the flags field of the head table? If so, I don't see any evidence that is touched during signing or verification.  

If you are referring to flags in the DSIG, the DSIG table is "removed" prior to computing the checksum and hash for verification or signing.

**Comment by ValZapod** [comment]
Okay

> the DSIG table is "removed" prior to computing

 What about the header of DSIG? At what point it is added? It is obvious that length of the PKCS#7 packet canot be added before, but the Version number of the DSIG table (0x00000001), number of signatures in the table (0x0001) and usFlag (named as such in https://docs.microsoft.com/en-us/typography/opentype/otspec180/dsig but not in https://docs.microsoft.com/en-us/typography/opentype/spec/dsig, another error) can be added before and so is format of the signature.

**Comment by paullinnerud** [comment]
The DSIG header is part of the DSIG and added to the font when the rest of the DSIG is added.

**Comment by ValZapod** [comment]
@PeterCon Well, now it is obvious that we need to change "Zero
out the file checksum in the head table." to "Recalculate and fill in file checksum in the head table for this stripped binary." and remove "Add the usFlag (reserved, set at 1 for now) to the stream of bytes" whatsover.

**Comment by ValZapod** [comment]
Oh, somebody just leaked all source code of windows XP (like 10 hours ago) and I am right now looking into ntgdi\fondrv\tt\mssipotf folder, in particular mssipotf.cpp.

What is this in function OTSIPObjectOTF::VerifyIndirectData?
```
        // Retrieve the flag field from the DSIG table.
        // These bytes will be used as part of the hash.
        usFlag = pDsigTable->GetFlag ();

        // Compute the hash value for the file and compare it to the hash
        // value in the SIP_INDIRECT_DATA structure.

```

**Comment by PeterCon** [comment]
@paullinnerud 

>It does not appear that the code zeros out the checksum in the 'head' table prior to calling CryptHashData().

Please clarify: do you mean the head.checksumAdjustment field?

**Comment by ValZapod** [comment]
You know you can just google it https://stackoverflow.com/questions/25727592/what-is-checksumadjustment-in-ttf-otf-head-table-used-for and all source code has leaked...

**Comment by PeterCon** [comment]
That may be the case. But as I've got many, many issues for the spec to work on in parallel, it makes much more sense to let Paul, who has already interpreted the source code, to clarify this for me than to spend my time ramping up on that code from scratch.

**Comment by ValZapod** [comment]
> this for me than to spend my time ramping up on that code from scratch

I am prety sure that usFlag is used in Windows XP source code to calculate md5 while he said it is not (in sipobotf.cpp). I also see very strange stuff in signature.cpp that RSA part is not even verified in Windows XP. That is from looking into your leaked code for 5 minutes. I can spend more time. ;)

**Comment by PeterCon** [comment]
Thanks. In the meantime, I will continue addressing comments on axis value tables within the STAT table, specifications for an in-preparation extension to the COLR table, processing of GPOS/GSUB nested lookups, and start making sense of complex issues regarding the scale for the 'opsz' design axis and interaction with CSS.

**Comment by HinTak** [comment]
@paullinnerud I think you basically confirmed my findings a few years ago (see earlier comment). I would suggest correcting the spec to match the de facto signing / verification implementation at this point in time, given so many fonts have been signed for two decades...

**Comment by ValZapod** [comment]
> think you basically confirmed my findings a few years ago

No, we did not. We found out that binary is stripped from DSIG and then head.checksumAdjustment field is recalculated and from that bitstream the digest is done (using md5 in 2020, crazy, though Donald Chinn did say that the functions will need to be changed to support sha256 and others).

**Comment by HinTak** [comment]
@ValZapod are looking at 20-year-old code. The new implementation in FontVal does all the newer digests, because newer version of the signing implementation signs with newer digest algorithms. sha1(or sha256? one of the sha* family) was quite popular, and used by Adobe's fonts a few years ago.

Anyway, please let the Microsoft folks respond. Please do not answer question (and wrongly...) on their behalf.

**Comment by PeterCon** [comment]
@ValZapod : I have a recollection that some changes (I don't know what) were made around the Windows 7 timeframe. So, I wouldn't necessarily assume that XP would provide definitive answers.

**Comment by HinTak** [comment]
This commit https://github.com/HinTak/Font-Validator/commit/8d23f52b8e9dedb1c53ad147f9c4bdfd81bbfc7c was created in Apr 2016 and made public in July 2017. It says "support all current valid arguments to HashAlgorithm.Create() - apparently Adobe starts to use SHA256" . So somehow, I have received reports/requests 4 years ago from somebody else that sha256-signed fonts from Adobe are out there.

**Comment by paullinnerud** [comment]
@peterCon Yes, I meant head.checksumAdjustment.
flags in the DSIG header indicates set to 0 however later it indicated set to 1. The constructor for the object in code sets to 1 and it's never set to anything else.

**Comment by paullinnerud** [comment]
mssipotf.dll had a major update in 2009 and smaller update in 2012. The last change in 2012 was to support SHA2 signatures according to the checkin log.

**Comment by ValZapod** [comment]
> The last change in 2012 was to support SHA2 signatures according to the checkin log.

Where? In RSA or in digest? I was talking about the digest.

**Comment by HinTak** [comment]
> mssipotf.dll had a major update in 2009 and smaller update in 2012. The last change in 2012 was to support SHA2 signatures according to the checkin log.

Cool - I think I have the 2009 dll's.

**Comment by PeterCon** [comment]
IIUC, this revision is what's needed. (The level 2 numbering is from GitHub's markdown, not quite what would be in the spec.) Confirmation from Paul and others requested:

>1. If there is an existing DSIG table in the font,
>    1. Remove <ins>the </ins>DSIG table from font.
>    2. Remove <ins>the </ins>DSIG table entry from sfnt [Table Directory](otff.md#tabledirectory).
>    3. Adjust table offsets as necessary.  
>   <del>4. Zero out the file checksum in the &#39;head&#39; table.</del>  
>   <del>5. Add the usFlag (reserved, set at 1 for now) to the stream of bytes</del>
>2. Hash the full stream of bytes using a secure one-way hash (such as MD5) to create the content digest.
>3. Create the PKCS#7 signature block using the content digest.
>4. Create a new DSIG table containing the signature block<ins>, and setting the flag field as desired</ins>.
>5. Add the DSIG table to the font, adjusting table offsets as necessary.
>6. Add a DSIG table entry to the sfnt Table Directory.
>7. Recalculate the checksum in the &#39;head&#39; table.

**Comment by PeterCon** [comment]
So,  this?

>1. If there is an existing DSIG table in the font,
>    1. Remove <ins>the </ins>DSIG table from font.
>    2. Remove <ins>the </ins>DSIG table entry from sfnt [Table Directory](otff.md#tabledirectory).
>    3. Adjust table offsets as necessary.  
>    4. <del>Zero out the file checksum in the &#39;head&#39; table.</del><ins>Recalculate checksumAdjustment in the &#39;head&#39; table.</ins>  
>   <del>5. Add the usFlag (reserved, set at 1 for now) to the stream of bytes</del>

**Comment by ValZapod** [comment]
5 should also still be there. See https://github.com/shaswata56/WindowsXP/blob/ef9fc49453d1788eadf64ece58451605d854e05a/windows/core/ntgdi/fondrv/tt/mssipotf/sign/sign.cpp#L108 usFlag is assigned to pbDsig, then https://github.com/shaswata56/WindowsXP/blob/ef9fc49453d1788eadf64ece58451605d854e05a/windows/core/ntgdi/fondrv/tt/mssipotf/sign/sign.cpp#L117 function HashTTFfile processes &pbHash

In HashTTFfile function, here https://github.com/shaswata56/WindowsXP/blob/ef9fc49453d1788eadf64ece58451605d854e05a/windows/core/ntgdi/fondrv/tt/mssipotf/sign/httffile.cpp#L483 we and I quote "Pump the extra bytes associated with the DSIG table into the hash function".

**Comment by HinTak** [comment]
> So, this?
> 
> > 1. If there is an existing DSIG table in the font,
> >    
> >    1. Remove the DSIG table from font.
> >    2. Remove the DSIG table entry from sfnt [Table Directory](otff.md#tabledirectory).
> >    3. Adjust table offsets as necessary.
> >    4. ~Zero out the file checksum in the 'head' table.~Recalculate checksumAdjustment in the 'head' table.
> >       ~5. Add the usFlag (reserved, set at 1 for now) to the stream of bytes~

Yes, this is basically what I having been suggesting for several years - remove the 4th item. Not sure about the 5th item either way - I think it was a provision for multiple / chained signatures. I have not seen it done.

**Comment by HinTak** [comment]
@PeterCon kind of beating a dead horse - https://github.com/MicrosoftDocs/typography-issues/issues/455#issuecomment-679192579 - I have been suggesting just removing the "zero out checksum..." item for years...

**Comment by paullinnerud** [comment]
The step "Add the usFlag (reserved, set at 1 for now) to the stream of bytes" should still be there but the flag is added to the hash regardless of if there is existing DSIG or not. It's also added to hash after file data is added. 

1. If there is an existing DSIG table in the font,
    a. Remove DSIG table from font.
    b.  Remove DSIG table entry from sfnt Table Directory.
    c.  Adjust table offsets as necessary. 
2. Hash the full stream of bytes using a secure one-way hash (such as MD5) to create the content digest.
3. Add the flags (uint16) (reserved, set at 1 for now) to the stream of bytes
4. Create the PKCS#7 signature block using the content digest.
5. Create a new DSIG table containing the signature block.
6. Add the DSIG table to the font, adjusting table offsets as necessary.
7. Add a DSIG table entry to the sfnt Table Directory.
8. Recalculate the checksum in the 'head' table.

**Comment by ValZapod** [comment]
> basically what I having been suggesting for several years - remove the 4th item

You do understand it does not remove the 4th item, right?

**Comment by ValZapod** [comment]
> * Hash the full stream of bytes using a secure one-way hash (such as MD5) to create the content digest.
> * Add the flags (uint16) (reserved, set at 1 for now) to the stream of bytes

You need to change places for those 2. Hash happens after! What is the point to say add usFlag, if it done after the hash? Hash md5 digest is the only thing that is signed by RSA + timestamp, so those buytes will be lost and you can change that flag while preserving the signature 100%! That is simply wrong.

**Comment by paullinnerud** [comment]
In the simplest case where the font does not already have a DSIG, the entire font image is sent to CryptHashData() and then the uint16 of the flag is passed to CryptHashData().

**Comment by paullinnerud** [comment]
So the wording needs to change to make it explicit that both are part of the hash?

**Comment by ValZapod** [comment]
> does not already have a DSIG, the entire font image is sent to CryptHashData() and then the uint16 of the flag is passed to CryptHashData().

Yes, what is your point? From https://docs.microsoft.com/en-us/windows/win32/api/wincrypt/nf-wincrypt-crypthashdata

```
hHash

Handle of the hash object.

pbData

A pointer to a buffer that contains the data to be added to the hash object.

```

**Comment by ValZapod** [comment]
> needs to change to make it explicit that both are part of the hash?

Yes.

**Comment by PeterCon** [comment]
Re `Add the flags (uint16) (reserved, set at 1 for now)`, the spec needs to be clear as to what `flags` refers to. 

If it is the DSIG.flags field, then bit 1 is _not_ reserved, and the current wording in the spec suggests that setting bit 1 is optional. Are we saying that setting bit 1 is _required_?

**Comment by paullinnerud** [comment]
Something like this.

1. If there is an existing DSIG table in the font,
    a. Remove DSIG table from font.
    b.  Remove DSIG table entry from sfnt Table Directory.
    c.  Adjust table offsets as necessary. 
2. Add font data to stream of bytes.
3. Add the flags (uint16) (reserved, set at 1 for now) to the stream of bytes
2. Hash the full stream of bytes using a secure one-way hash (such as MD5) to create the content digest.
4. Create the PKCS#7 signature block using the content digest.
5. Create a new DSIG table containing the signature block.
6. Add the DSIG table to the font, adjusting table offsets as necessary.
7. Add a DSIG table entry to the sfnt Table Directory.
8. Recalculate the checksum in the 'head' table.

**Comment by paullinnerud** [comment]
BTW, this area of the code has changed considerably from the original version.

**Comment by paullinnerud** [comment]
The constructor for the object that manages flags field of DSIG sets it to 1 and it is never changed.

**Comment by ValZapod** [comment]
> 1. b.  Remove DSIG table entry from sfnt Table Directory.
>    c.  Adjust table offsets as necessary.

You also forgot about 

> The reason this does not break the signature is when the signature is verified, an in memory image is created of the font without the DSIG and the checksum is recalculated before the data is hashed.

Now:


> object that manages flags field of DSIG sets it to 1 and it is never changed.

Because only 1 is allowed. "permission flags Bit 0: cannot be resigned Bits 1-7: Reserved (Set to 0)"
As I said,

> usFlag (named as such in https://docs.microsoft.com/en-us/typography/opentype/otspec180/dsig but not in https://docs.microsoft.com/en-us/typography/opentype/spec/dsig, another error)

#

**Comment by PeterCon** [comment]
>The constructor for the object that manages flags field of DSIG sets it to 1 and it is never changed.

But that is one implementation. For the DSIG table to work, is it always necessary for the flag to be set to 1? Or could a font have flags set to 0?

**Comment by ValZapod** [comment]
> Or could a font have flags set to 0?

Why did you chage usFlag to just flags? Or who was it... It can, moreover they are copied here https://github.com/shaswata56/WindowsXP/blob/ef9fc49453d1788eadf64ece58451605d854e05a/windows/core/ntgdi/fondrv/tt/mssipotf/sign/sign.cpp#L97
Now he just said that it is only 1 in newer versions... so...

**Comment by PeterCon** [comment]
>Why did you chage usFlag to just flags? 

Field names in several tables have been revised in prior versions for consistency in naming conventions.

**Comment by ValZapod** [comment]
"If this bit is set to zero (0) the font may have a signature applied over the existing digital signature(s)."

That like with Nvidia drivers dlls, tney have 1 sha1 and one sha2 signature, both have 3 certificates of RSA PKI with 3 certificates for timestamp. That is 12 certificates, all different (they are crazy, I know).

**Comment by HinTak** [comment]
@PeterCon I am okay with what @paullinnerud suggests. As for the value of `usFlag`, I checked that it was a placeholder for concatenating signatures (for future implementations), zero allowing others to counter-sign until the last person who signs, sets it to 1. After 20 years, no signing implementation offers the functionality of counter-signing, and no verification implementation checks counter-signing, so the `usFlag` is effectively a reserved value of 1. @paullinnerud is saying that the MS implementation behaves always to write a "last signature", and forbid further counter-signing.

**Comment by paullinnerud** [comment]
This need to be there too "The reason this does not break the signature is when the signature is verified, an in memory image is created of the font without the DSIG and the checksum is recalculated before the data is hashed."

**Comment by HinTak** [comment]
I am kind-of against adding any description about memory models as it is an implementation details, and does not need to go into a specification. Keep things simple.

**Comment by ValZapod** [comment]
It is time for a pull request.

Who is going to submit this info to ISO body?

**Comment by PeterCon** [comment]
Once I get clarity on the details, I will prepare a revision for the OT spec, and present the proposed revision here for review. Later it will get submitted to ISO for their consideration.

**Comment by PeterCon** [comment]
>And it also seems that @paullinnerud is suggesting the verification side of MS implementation does not accept chained/concatenating/counter-signing.

OK that seems to suggest that bit 0 in the flags field should not be optional but is always expected to be set.

**Comment by ValZapod** [comment]
If there is an existing DSIG table in the font,
a. Remove DSIG table from font.
b. Remove DSIG table entry from sfnt Table Directory.
c. Adjust table offsets as necessary.
d. Recalculate checksumAdjustment in the 'head' table.
Add font data to stream of bytes.
Add the flags (uint16) (set at 1 for now as no crossign implementation) to the stream of bytes
Hash the full stream of bytes using a secure one-way hash (such as MD5 or SHA256) to create the content digest.
Create the PKCS#7 signature block using the content digest (block in many cases contains timestamping to make the signature valid before the decades long timestamping certificate will expire).
Create a new DSIG table containing the signature block.
Add the DSIG table to the font, adjusting table offsets as necessary.
Add a DSIG table entry to the sfnt Table Directory.
Recalculate checksumAdjustment in the 'head' table.

**Comment by HinTak** [comment]
> > And it also seems that @paullinnerud is suggesting the verification side of MS implementation does not accept chained/concatenating/counter-signing.
> 
> OK that seems to suggest that bit 0 in the flags field should not be optional but is always expected to be set.

Yes, it is reserved as 1 for now, until such times as somebody wants to implement counter-signing, and needs to define how counter-signing should work. This is perhaps similar to how the higher unicode range bits in the OS/2 table - they takes on a default and reserved value of 0 until somebody defines what 1 means. In this case, "1" means "this is the final signature". Nobody has done a "intermediate signature" signing, and the verifying always expects a final signature.

**Comment by ValZapod** [comment]
> About ttc header, the bulk of many old Microsoft shipped ttc also have a version 1 header rather than version 2. So that's another issue where the De facto signing implementation (does not reject v1 header) does not match the spec (which only states v2).

BTW, @HinTak, what does that mean? TTC is not supported by Windows, are they not? I understand you get rid of atmfd.dll, but TTC was never supported at all, not like with postscript fonts.

**Comment by HinTak** [comment]
Windows have been shipping a few ttc's since about 20 years ago. And they are signed, too.

**Comment by ValZapod** [comment]
> Windows have been shipping a few ttc's since about 20 years ago. And they are signed, too.

Oh, indeed. Press win and type C:\Windows\Fonts it will actually open the real thing that is under "The Interface". 18 ttc files. Then why they are not supported for instalation the last time I checked?

**Comment by PeterCon** [comment]
It's been possible to install TTCs in Windows since TTCs were first defined.

**Comment by PeterCon** [comment]
OK, I've drafted revisions. There are enough changes that I'm attaching a PDF for the entire chapter showing the revisions.

I did not mention timestamps since, IIUC, that would be a detail in the signature block covered by the PKCS #9 spec.

As the discussion brought up TTCs, I took a look at the TTC section toward the end and noticed it didn't give instructions about creating a DSIG table for a TTC. So, I listed what I think (IIUC) would be the appropriate adjustments to the steps discussed in the case of a TTC.

Please review.

[dsig_184_draft_201001.pdf](https://github.com/MicrosoftDocs/typography-issues/files/5315458/dsig_184_draft_201001.pdf)

**Comment by PeterCon** [comment]
Oh, and I also updated the references for PKCS #7/#9: the current links no longer work.

**Comment by HinTak** [comment]
@PeterCon :

- there is no need to remove "permission flags" in the description column of the header table . Removing it would look odd, given there is still a whole paragraph below about it.

- I would suggest using one of the urls under ietf.org instead of rfc-editor.org for pkcs 7/9. Looks more official. Strangely I have those ftp files in my hard drive (I tend to do downloads with mirroring tools, to keep the original urls) so they were available as written in Sept 2015.

- Argh, I see there is some confusion with the permission field in the signature header, and the two bytes which is appended to the end of the file stream during hashing. Both are 0x0001 and both are called `*flags` but have different meanings. One is a collection of permission bits, which nobody has bothered with implementing or verifying other possible values. The other is just two arbitrary bytes apend to the end of input data, to make the hashing more interesting (... ask your colleagues to explain why those two bytes and not something else...) . Perhaps saying "Append two bytes, 0x00 and 0x01 ..." is better than what is at item 2.

- perhaps a reference to ttc header sub-section of Font collection section? I think that section also needs a little clarification: it says "if a digital signature is used, the DSIG table for the file must be the last table in the TTC file". What it means, is that, for TTC, the DSIG table is appended to the TTC font file after signing, and the TTC header's dsigOffset field has a value somewhat near the end of file (minus size of the DSIG table). i.e. For. Signed TTC, the DSIG table is always located at the very ending portion of file. You can try rewriting my words on this in a nicer way.

I have no strong opinion, one way or the other, on the other changes.

**Comment by HinTak** [comment]
The "...last table..." part is "by content location", or "by physical location".

**Comment by PeterConstable** [comment]
>I would suggest using one of the urls under ietf.org instead of rfc-editor.org for pkcs 7/9. Looks more official. 

See <https://www.rfc-editor.org/ref-example/> for Internet Society guidance on referencing RFCs: URLs are of the form `http://www.rfc-editor.org/info/rfc#`.

**Comment by PeterConstable** [comment]
>there is no need to remove "permission flags" in the description column of the header table . Removing it would look odd, given there is still a whole paragraph below about it.

I missed that paragraph. If, in fact, bit 0 of the DSIG flags field is going to be required—which is what I was given to understand from the discussion—, then that paragraph is no longer applicable and should be removed.


>Argh, I see there is some confusion with the permission field in the signature header, and the two bytes which is appended to the end of the file stream during hashing. Both are 0x0001 and both are called *flags but have different meanings.

As I said earlier, we need to be clear what "flags" being added to the stream refers to. Nobody gave me any clarification, so I took it to mean the same as the only "flags" clearly in the context, namely the DSIG flags field.

If the bytes added to the stream are just that, bytes, then they're not flags; they're just arbitrary bytes.

I appreciate clarification on these points. Thanks.

**Comment by HinTak** [comment]
@PeterConstable sorry, I have switched off and blocked notifications from non-Microsoft staffers, and I also categorically refuse to read comments and observations on stolen source code. I see now that I am confused because there are no non-final signatures out there in the wild. A non-final signature is generated by hashing differently, with a different permission flag appended. In that case I have a history question, perhaps for @paullinnerud or @Gr3gH : what is or was the use-case / business case for a non-final signature? Given that font vendors almost always want to ship and release fonts as final-signed, if signed at all.

**Comment by ValZapod** [comment]
> If the bytes added to the stream are just that, bytes, then they're not flags; they're just arbitrary bytes.

They are bytes of flags, previously usFlag.

BTW, I do not understand how counter-signing even is supposed to work. Changing the flag from 0 to 1 will change the first signature, will it not?

**Comment by PeterCon** [comment]
>there is no need to remove "permission flags" in the description column of the header table 

OK, I got further clarification from Paul: the "flags" appended to the stream are the same flags to be written to DSIG.flags in step 5, or (when validating a font's signature) that are read from the font's DSIG.

So, here's an updated draft in which I
* restore the permission flags description
* revise step 2 (flags value is being selected before the DSIG is created)
* revise the paragraph after the steps to clarify the validation process wrt flags

[dsig_184_draft_201002_diff-from-183.pdf](https://github.com/MicrosoftDocs/typography-issues/files/5320695/dsig_184_draft_201002_diff-from-183.pdf)

**Comment by PeterCon** [comment]
>perhaps a reference to ttc header sub-section of Font collection section? I think that section also needs a little clarification:

OK, I should have done this before uploading the draft in the previous comment: I've added the following at the end of the section on signatures for TTCs:

><ins>See the [TTC Header](otff.md#ttc-header) section of the OpenType Font File chapter for related information.</ins>

And in that other section of the OTFF chapter, I've revised this paragraph:

>If a digital signature is used, the DSIG table for the file must be <del>the last table in the TTC file</del><ins>located at the end of the TTC file, following any other font tables</ins>. Signatures in a TTC file are expected to be Format 1 signatures.

**Comment by HinTak** [comment]
@PeterCon 

> OK, I got further clarification from Paul: the "flags" ...
> * revise the paragraph after the steps to clarify the validation process wrt flags
> 

Thanks a lot for the clarification and revisions. It is all good to me now. I have not seen a non-final signature in the wild, but I'll know one what I see one in the future. It is also great that you have condensed my rambling on the ttc header into a neat and clear change.

FWIW, the code in
https://github.com/HinTak/Font-Validator/tree/master/DSIGInfo is an interface-compatible drop-in replacement to mssipotf (its dotnet exposure), somewhat incomplete, but cross-platform and works on linux/mac os too. It was written by just looking how shipping fonts' signatures look like, assuming officially released fonts have valid signatures. So I have not seen a non-final signature. When FontVal was open-sourced by Microsoft, mssipotf (which was always bundled with FontVal releases) was not opened. I don't look at source code I do not legally have. It is good to see details clarified.

**Comment by PeterCon** [comment]
With no further comments suggestion corrections needed, I'm assuming this is resolved, so closing.

**Comment by ValZapod** [comment]
> Later it will get submitted to ISO for their consideration.

Any progress on ISO/IEC standard? Will it be in https://www.iso.org/standard/82487.html (AWI AMD 2)?

**Comment by vlevantovsky** [comment]
> Any progress on ISO/IEC standard?

The work on AMD2 is currently underway. Yes, this update can definitely be part of the new amendment - I opened a [new issue 45](https://github.com/MPEGGroup/OpenFontFormat/issues/45) to address it.
